File: translator-baidu-api/src/main/java/com/manerfan/translator/api/baidu/TranslatorManager.java
/*
 * ManerFan(http://manerfan.com). All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.manerfan.translator.api.baidu;

import com.google.common.base.Joiner;
import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListeningExecutorService;
import com.google.common.util.concurrent.MoreExecutors;
import com.manerfan.translator.api.baidu.body.BaiduResponseBody;
import com.manerfan.translator.api.baidu.body.BaiduTransBody;
import com.manerfan.translator.api.baidu.filters.FilterContext;
import com.manerfan.translator.api.baidu.filters.TransFilter;
import com.manerfan.translator.api.sbd.SentenceSegmentation;
import com.manerfan.translator.jpa.repositories.StatisticsRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.CollectionUtils;
import org.springframework.util.StringUtils;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * @author manerfan
 * @date 2017/10/16
 */

public class TranslatorManager extends BaiduTransUtil {
    Logger logger = LoggerFactory.getLogger(TranslatorManager.class);

    Pattern preBlankPattern = Pattern.compile("^[\\s]+");

    @Autowired
    SentenceSegmentation sentenceSegmentation;

    @Autowired
    StatisticsRepository statisticsRepository;

    ListeningExecutorService executorService;

    List<TransFilter> filters = new LinkedList<>();

    public TranslatorManager(ExecutorService executorService) {
        this.executorService = MoreExecutors.listeningDecorator(executorService);
    }

    public TranslatorManager addFilter(TransFilter transFilter) {
        filters.add(transFilter);
        return this;
    }

    /**
     * 短文本翻译
     *
     * @param q    required，要翻译的文本
     * @param from optional，翻译源语言，可设置为auto
     * @param to   optional，译文语言，缺省为zh
     *
     * @return 译文
     *
     * @throws Exception
     */
    public BaiduResponseBody transText(String q, String from, String to) throws Exception {
        Preconditions.checkArgument(StringUtils.hasText(q), "q should not be empty");

        FilterContext context = FilterContext.<String>newInstance();

        /* 预处理 */
        for (TransFilter filter : filters) {
            q = filter.pre(q, context);
        }

        BaiduResponseBody response = transReq(q, from, to);

        /* 后处理 */
        for (TransFilter filter : filters) {
            response = filter.post(response, context);
        }

        statisticsRepository.increaseTextNum();
        statisticsRepository.increaseBytesNum(q.getBytes().length);

        return response;
    }

    /**
     * 批量短文本翻译
     *
     * @param qs   required，要翻译的文本
     * @param from optional，翻译源语言
     * @param to   optional，译文语言
     *
     * @return 译文
     *
     * @throws Exception
     * @see TranslatorManager#transText(String, String, String)
     */
    private List<BaiduResponseBody> transTextArray(Collection<String> qs, String from, String to) throws Exception {
        Preconditions.checkArgument(!CollectionUtils.isEmpty(qs), "qs should not be empty");

        return Futures.allAsList(qs.stream().map((q) -> executorService.submit(
                () -> transText(q, from, to))
        ).collect(Collectors.toList())).get();
    }

    /**
     * 将长文本拆分为短文本
     *
     * @param sentence 长文本
     *
     * @return 短文本
     */
    private List<String> fixedToSentences(String sentence) {
        SentenceSegmentation.Sentences segment = sentenceSegmentation.segment(sentence);

        int l = 0;
        StringBuilder sb = new StringBuilder();
        List<String> sentences = Lists.newLinkedList();
        while (segment.hasNext()) {
            String s = segment.next();
            int len = s.getBytes().length;

            if (l + len > limit) {
                sentences.add(sb.toString());
                sb = new StringBuilder();
                l = 0;
            }

            l += len;
            sb.append(s);
        }
        sentences.add(sb.toString());

        return sentences;
    }

    /**
     * 长文本翻译（忽略回车符）
     *
     * @param s    required，要翻译的文本
     * @param from required，翻译源语言
     * @param to   required，译文语言
     *
     * @return 译文
     *
     * @throws Exception
     */
    BaiduResponseBody transSection(String s, String from, String to) throws Exception {
        Preconditions.checkArgument(StringUtils.hasText(s), "s should not be empty");
        Preconditions.checkArgument(StringUtils.hasText(from), "from should not be empty");
        Preconditions.checkArgument(StringUtils.hasText(to), "to should not be empty");

        List<String> sentences;
        if (s.getBytes().length > limit) {
            sentences = fixedToSentences(s);
        } else {
            sentences = Lists.newArrayList(s);
        }

        List<BaiduResponseBody> bodies = transTextArray(sentences, from, to);
        String dst = Joiner.on("").join(bodies.stream().map((body) -> body.getDst()).collect(Collectors.toList()));

        return new BaiduResponseBody(from, to, Lists.newArrayList(new BaiduTransBody(s, dst)));
    }

    /**
     * 批量翻译句子/段落<br/>
     * (忽略每一条子数据中的回车符)
     *
     * @param sentences required，要翻译的句子/段落
     * @param from      required，翻译源语言
     * @param to        required，译文语言
     *
     * @return 顺序相对应的翻译结果
     *
     * @throws Exception
     */
    public List<String> transSentences(List<String> sentences, String from, String to) throws Exception {
        if (CollectionUtils.isEmpty(sentences)) {
            return Lists.newArrayList();
        }

        List<SentenceWrapper> wrappers = wrapSentences(sentences);
        List<List<SentenceWrapper>> groupWarppers = groupWrapperList(wrappers);
        groupWarppers.parallelStream().forEach((groupWrapper) -> {
            try {
                if (groupWrapper.size() > 1) {
                    List<BaiduResponseBody> dsts = transTextArray(
                            groupWrapper.stream().map((wrapper) -> wrapper.src).collect(Collectors.toList()), from, to);

                    int[] idx = {0};
                    groupWrapper.stream().forEach((wrapper) -> wrapper.dst = dsts.get(idx[0]++).getDst());

                } else {
                    groupWrapper.get(0).dst = transSection(groupWrapper.get(0).src, from, to).getDst();
                }
            } catch (Exception e) {
                logger.error("TransSentence Error!", e);
            }
        });

        return wrappers.stream().map(SentenceWrapper::dstString).collect(Collectors.toList());
    }

    private List<List<SentenceWrapper>> groupWrapperList(List<SentenceWrapper> wrappers) {
        List<List<SentenceWrapper>> groupWrappers = Lists.newLinkedList();

        int l = 0;
        List<SentenceWrapper> groupWrapper = Lists.newLinkedList();
        Iterator<SentenceWrapper> wrapperIterator = wrappers.iterator();
        while (wrapperIterator.hasNext()) {
            SentenceWrapper wrapper = wrapperIterator.next();
            if (!wrapper.srcHasText()) {
                // 不含有效字符的略过
                continue;
            }

            int len = wrapper.srcBytesLen();

            if (l + len > limit) {
                groupWrappers.add(groupWrapper);
                groupWrapper = Lists.newLinkedList();
                l = 0;
            }

            l += len;
            groupWrapper.add(wrapper);
        }
        groupWrappers.add(groupWrapper);

        return groupWrappers;
    }

    private List<SentenceWrapper> wrapSentences(List<String> sentences) {
        return sentences.stream().map(SentenceWrapper::new).collect(Collectors.toList());
    }

    private class SentenceWrapper {
        String prefix = "";
        String src;
        String dst;

        public SentenceWrapper(String src) {
            Matcher matcher = preBlankPattern.matcher(src);
            if (matcher.find()) {
                prefix = matcher.group();
                src = matcher.replaceFirst("");
            }

            this.src = src;
            this.dst = this.src;
        }

        boolean srcHasText() {
            return StringUtils.hasText(src);
        }

        int srcBytesLen() {
            return src.getBytes().length;
        }

        public String dstString() {
            return prefix + dst;
        }
    }
}

File: translator-baidu-api/src/main/kotlin/com/manerfan/translator/api/baidu/TranslatorManager.kt
package com.manerfan.translator.api.baidu

import com.google.common.base.Preconditions
import com.google.common.collect.Lists
import com.manerfan.translator.BusinessException
import com.manerfan.translator.ErrorCode
import com.manerfan.translator.api.sbd.SentenceSegmentation
import com.manerfan.translator.data.StatisticsRepository
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.runBlocking
import org.apache.commons.pool2.BasePooledObjectFactory
import org.apache.commons.pool2.impl.DefaultPooledObject
import org.apache.commons.pool2.impl.GenericObjectPool
import org.apache.commons.pool2.impl.GenericObjectPoolConfig
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.math.BigInteger
import java.security.MessageDigest

/**
 * @author manerfan
 * @date 2017/12/6
 */

@Component
class Md5 {
    private lateinit var md5DigestPool: GenericObjectPool<MessageDigest>

    fun encrypt(str: String): String {
        val md5 = md5DigestPool.borrowObject()
        try {
            md5.reset()
            md5.update(str.toByteArray())
            return BigInteger(1, md5.digest()).toString(16).padStart(32, '0')
        } finally {
            md5DigestPool.returnObject(md5)
        }
    }

    private fun md5DigestPoolFactory(): BasePooledObjectFactory<MessageDigest> {
        return object : BasePooledObjectFactory<MessageDigest>() {
            override fun create() = MessageDigest.getInstance("MD5")
            override fun wrap(messageDigest: MessageDigest) = DefaultPooledObject(messageDigest)
        }
    }

    init {
        val poolConfig = GenericObjectPoolConfig()

        /* 设置池中最大idle，若idle数量大于此值，则会清理多余idle */
        poolConfig.maxIdle = 100
        /* 设置多久没有borrow则设置为idle */
        poolConfig.minEvictableIdleTimeMillis = (5 * 60 * 1000).toLong()

        /* 设置池中最小idle，若idle数量小于此值，则在Evictor定时器中会自动创建idle */
        poolConfig.minIdle = 50
        /* 设置Evictor定时器周期并启动定时器 */
        poolConfig.timeBetweenEvictionRunsMillis = (30 * 1000).toLong()

        /* 设置池中最大数量，若达到上限时borrow，则阻塞 */
        poolConfig.maxTotal = 500

        /* 调用者最大阻塞的时间 */
        poolConfig.maxWaitMillis = (5 * 1000).toLong()

        md5DigestPool = GenericObjectPool(md5DigestPoolFactory(), poolConfig)
        // 初始化池中idle个数到minIdle，提前create，免得在使用时再创建
        // BaseGenericObjectPool.Evictor定时器会定时执行ensureMinIdle，确保idle keypare个数可以达到minIdle
        md5DigestPool.preparePool()
    }
}

@Component
class TranslatorManager(
        val md5: Md5,
        val restService: RestService,
        val sentenceSegmentation: SentenceSegmentation,
        val filters: List<TranslatorFilter<String>>
) {
    private val logger = LoggerFactory.getLogger(TranslatorManager::class.java)

    @Value("\${baidu.trans.app.id}")
    lateinit var appId: String

    @Value("\${baidu.trans.app.secret}")
    private lateinit var appSecret: String

    @Value("\${baidu.trans.limit:5000}")
    private var limit: Int = 0

    @Autowired
    lateinit var statisticsRepository: StatisticsRepository

    /**
     * 生成签名
     */
    private fun sign(q: String, salt: Long) = md5.encrypt("$appId$q$salt$appSecret").padStart(32, '0')

    /**
     * 翻译
     */
    private fun translate(q: String, from: String = "auto", to: String = "zh"): BaiduResponseBody {
        Preconditions.checkArgument(q.length <= limit, "q is too long (${q.length}) than $limit bytes")

        val salt = System.currentTimeMillis()
        var resp = restService.translate(q, from, to, appId, salt, sign(q, salt)).execute()
        if (200 != resp.code()) {
            throw BusinessException(ErrorCode.REST_SERVICE_ERROR, resp.body()?.toString())
        }

        var body = resp.body() ?: throw BusinessException(ErrorCode.REST_SERVICE_ERROR, "Baidu Translator Api Response Empty!")
        body.error_code?.let { throw BusinessException(ErrorCode.REST_SERVICE_ERROR, resp.body()?.toString()) }

        return body
    }

    /**
     * 短文本翻译
     */
    fun transText(q: String, from: String, to: String): BaiduResponseBody {
        var query = q
        var context = FilterContext<String>()

        /* 预处理 */
        filters.forEach { query = it.pre(query, context) }

        /* 翻译 */
        var response = translate(query, from, to)

        /* 后处理 */
        filters.forEach { response = it.post(response, context) }

        statisticsRepository.increaseTextTimes()
        statisticsRepository.increaseBytesNum(q.length.toLong())

        return response
    }

    /**
     * 批量短文本翻译
     */
    fun transTextArray(qs: Collection<String>, from: String, to: String): List<BaiduResponseBody> {
        return runBlocking { qs.map { async(CommonPool) { transText(it, from, to) } }.map { it.await() } }
    }

    /**
     * 将长文本拆分为短文本列表
     */
    private fun fixedToSentences(sentence: String): List<String> {
        var segments = sentenceSegmentation.segment(sentence)

        var sb = StringBuilder()
        var sentences = Lists.newLinkedList<String>()

        for (segment in segments) {
            if (sb.length + segment.length > limit) {
                sentences.add(sb.toString())
                sb = StringBuilder()
            }
            sb.append(segment)
        }
        sentences.add(sb.toString())

        return sentences
    }

    /**
     * 长文本翻译（忽略回车符）
     */
    fun transSection(s: String, from: String, to: String): BaiduResponseBody {
        var dst = transTextArray(fixedToSentences(s), from, to)
                .flatMap { it.trans_result.orEmpty() }.joinToString("") { it.dst }

        return BaiduResponseBody(from, to, listOf(BaiduTransBody(s, dst)))
    }

    /**
     * 批量翻译句子/段落（忽略每一条子数据中的回车符）
     */
    fun transSectionArray(sentences: Collection<String>, from: String, to: String): List<String> {
        return groupWrapperList(sentences).onEach {
            when {
                it.size > 1 -> transTextArray(it.map(SentenceWrapper::src), from, to).forEachIndexed { index, baiduResponseBody ->
                    it[index].dst = baiduResponseBody.trans_result.orEmpty().joinToString("") { it.dst }
                }
                else -> it[0].dst = transSection(it[0].src, from, to).trans_result.orEmpty().joinToString("") { it.dst }
            }
        }.flatMap { it.map(SentenceWrapper::dst) }
    }

    private fun groupWrapperList(sentences: Collection<String>): List<List<SentenceWrapper>> {
        var groupWrappers = mutableListOf<List<SentenceWrapper>>()
        var wrappers = mutableListOf<SentenceWrapper>()

        var len = 0
        sentences.filter(String::isNotBlank).forEach {
            if (len + it.length > limit) {
                groupWrappers.add(wrappers)
                wrappers = mutableListOf()
                len = 0
            }

            len += it.length
            wrappers.add(SentenceWrapper(it))
        }
        groupWrappers.add(wrappers)
        return groupWrappers
    }

    private inner class SentenceWrapper(var src: String) {
        private var preBlankPattern = "^[\\s]+".toPattern()

        var prefix = ""
        var dst: String
            get() = prefix + field

        init {
            var matcher = preBlankPattern.matcher(src)
            if (matcher.find()) {
                prefix = matcher.group()
                src = matcher.replaceFirst("")
            }
            dst = src
        }
    }
}
